### ISSUE TITLE
PROJECT-KAGYAを24GB単一GPUで完結させるHardware Contractを定義する

### ISSUE THREAD

**Issue by upiscium** [upiscium/PROJECT-KAGYA#272]
## 背景

PROJECT-KAGYAの初期設計では、実運用時に必要なGPU VRAM容量やmulti-GPU前提の有無は明示的に固定されていなかった。

今後の実装・学習機構・System One統合を進めるにあたり、ハードウェア要件が際限なく拡大することを避けたい。

そのため、PROJECT-KAGYAの基本Hardware Contractとして、

> **NVIDIA GPU 1枚 / VRAM 24 GiBで、通常推論・内部状態処理・sleep-time learning・adapter evaluationまで完結可能であること**

を正式な設計目標として追加する。

本IssueはSystem One/Jev統合そのものではなく、PROJECT-KAGYA全体に対するhardware/resource constraintを定義することを目的とする。

関連:
- #271 KAGYA専用0.6B System One (Jev) を24GB単一GPU構成に統合する

## 目的

以下をPROJECT-KAGYAの明示的な設計制約とする。

- 単一NVIDIA GPUで動作する
- VRAM上限は24 GiBを基準とする
- 通常運用時にmulti-GPUを必須としない
- QLoRA等のsleep-time learningも同一GPUで実行可能にする
- 推論系と学習系は時間分割し、必要に応じてmodel unload/reloadを行う
- 24 GiBを超える要求が出た場合、まずmodel/runtime/training profileを最適化する
- GPU追加を前提に問題を解決しない

## 想定対象GPU

Reference hardware:

- NVIDIA RTX 3090 24GB

ただしRTX 3090固有実装にはしない。

要件は、

```text
CUDA-capable NVIDIA GPU
VRAM >= 24 GiB
```

程度のcapability contractとして扱う。

## AWAKE時の要件

通常運用時は以下を同一24GB GPU内で成立させる。

- Gemma 4 E4B base model
- NF4等の量子化
- ACTIVE LoRA
- KAGYA通常推論
- prediction error / loss calculation
- memory retrievalとruntime処理
- #271で導入予定の0.6B級System Oneを含む場合も24GB内に収める
- 必要なKV cache / CUDA workspaceを含めてOOMしない

初期acceptance target:

```text
peak allocated/reserved VRAM < 23 GiB
```

1 GiB程度はruntime variation / fragmentation / driver差分の余裕として残す。

## SLEEP時の要件

SLEEP時はAWAKE用modelを同時常駐させ続けることを要求しない。

想定flow:

```text
AWAKE
  ↓
new interaction admission stop
  ↓
current turn drain
  ↓
state / memory flush
  ↓
inference subprocess stop
  ↓
VRAM release
  ↓
SLEEP workloads
  ├─ memory consolidation
  ├─ dream dataset generation
  ├─ Gemma 4 E4B QLoRA
  ├─ candidate adapter evaluation
  └─ optional System One training
  ↓
training/evaluation subprocess stop
  ↓
AWAKE inference reload
  ↓
service resume
```

同一process内の `del model` / `torch.cuda.empty_cache()` だけに依存せず、subprocess terminationをVRAM cleanup boundaryとして利用できる構造を優先する。

## QLoRA 24GB profile

初期QLoRA profileは24GB制約を優先して保守的に設定する。

- 4-bit NF4
- double quantization
- BF16 compute
- LoRA rank 16を初期値とする
- per-device batch size 1
- gradient accumulation
- gradient checkpointing
- `use_cache = false`
- paged / memory-efficient optimizerを優先
- initial max sequence lengthは2048程度
- sequence lengthは実測後に4096 / 8192へ段階的に拡張
- multimodal QLoRAは初期scope外

## VRAM Budgetを明示する

各subsystemは「動いた」だけではなく、VRAM budgetを測定・記録する。

最低限:

- base model weights
- quantization overhead
- LoRA parameters
- KV cache
- activations
- optimizer states
- gradients
- CUDA/runtime workspace
- System One model
- fragmentation / safety margin

実測値をhardware acceptance結果として残す。

## Runtime policy

24GB制約を維持するため、以下を許容する。

- model quantization
- model unload / reload
- subprocess分離
- batch size削減
- gradient accumulation
- gradient checkpointing
- sequence length制限
- adapter-only training
- optional componentのSLEEP時unload
- AWAKE / SLEEP間の時間分割

逆に、以下をデフォルト解決策にしない。

- model parallel
- tensor parallel
- multi-GPU必須化
- GPU間offload前提
- 48GB / 80GB級GPUを前提とする設計
- 別GPUをtraining専用に要求する設計

## Degradation policy

24GB内に収まらない新機能を追加する場合は、即座にhardware requirementを引き上げず、以下の順で再検討する。

1. 不要な同時常駐を減らせないか
2. componentをAWAKE/SLEEPで時間分割できないか
3. quantization / checkpointingで削減できないか
4. sequence / batch / cache policyを調整できないか
5. smaller model / adapter / specialized modelへ置換できないか
6. それでも不可能な場合のみhardware contract変更を別Issueで審議する

24GB超過は暗黙に許容しない。

## PROJECT-KAGYA only

このHardware Contractを満たすために、KAGYA node内へ以下を必須依存として持ち込まない。

- homelab-llm-platform worker runtime
- 外部GPU scheduler
- 外部training service
- 外部model-serving daemon

PROJECT-KAGYA repository内のコードと通常のPython/CUDA依存だけで再構築可能な状態を維持する。

Home Lab基盤との連携がある場合も、KAGYAの内部実行成立条件にはしない。

## Hardware Acceptance Test

最低限、実GPU上で以下を確認する。

### AWAKE

- [ ] PROJECT-KAGYAをcold startできる
- [ ] Gemma 4 E4Bを24GB GPUへロードできる
- [ ] ACTIVE LoRAをattachできる
- [ ] normal generationが成功する
- [ ] loss calculationが成功する
- [ ] memory-backed turnが成功する
- [ ] System One導入後も同時常駐できる
- [ ] peak VRAM < 23 GiB

### SLEEP

- [ ] AWAKE runtimeを安全に停止できる
- [ ] VRAMがtraining開始可能な状態まで解放される
- [ ] dream datasetを生成できる
- [ ] Gemma 4 E4B QLoRAを24GB内で実行できる
- [ ] candidate adapter artifactを生成できる
- [ ] adapter evaluationを実行できる
- [ ] training/evaluation runtimeを停止できる
- [ ] AWAKE runtimeを再ロードできる
- [ ] prior canonical state / memoryを保持したまま運用再開できる

## 非目標

- 24GB GPUで最大throughputを達成すること
- AWAKE中にQLoRAを並行実行すること
- multi-GPU trainingを禁止すること

multi-GPUは研究・実験用途として将来追加してもよいが、PROJECT-KAGYAの標準運用要件にはしない。

## 完了条件

- [ ] 24GB single-GPUをPROJECT-KAGYAの正式Hardware Contractとしてdocumentationへ追加
- [ ] AWAKE/SLEEP resource lifecycleを定義
- [ ] VRAM measurement方法を定義
- [ ] 24GB向けQLoRA profileを定義
- [ ] hardware acceptance procedureを追加
- [ ] RTX 3090 24GB級実機でacceptanceを通す
- [ ] 24GB超過時のdegradation policyをdocumentationへ追加
- [ ] #271のSystem One導入後もこのcontractを維持する
